v2x_ros_driver/src/v2x_radio_client.cpp

```diff
@@ -168,29 +169,124 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
+
             size_t start_index = i;
             size_t end_index = i + 2 + len + len_bytes; // includes 2 msgID bytes before message body
-            //this constructor has range [first, last) hence the + 1
+            // this constructor has range [first, last) hence the + 1
             std::vector<uint8_t> msg_vec(entry.begin() + start_index, entry.begin() + end_index);
-            onMessageReceived(msg_vec, msg_id);
-            break;
+
+            // Second check for a valid message. Checks MessageFrame length field against actual payload size.
+            if (!isValidMsgSize(msg_vec, start_index, end_index, entry))
+            {
+                RCLCPP_WARN_STREAM(logger_, "Size in possible MessageFrame does not match actual data size. Checking rest of data.");
+                continue;
+            }
+            else
+            {
+                // Last check for a valid message. WSA uses same format as WAVE Short Message (WSM) and precurses the WSM in some cases.
+                // Make sure WSA is not accidentally detected before actual message. This is done by checking detected msg_id against list of PSIDs.
+                if (!isValidPSID(std::to_string(msg_id)))
+                {
+                    printVectorHelper(msg_vec);
+                    onMessageReceived(msg_vec, msg_id);
+                    break;
+                }
+                else
+                {
+                    RCLCPP_WARN_STREAM(logger_, "PSID found, parsing rest of data for MessageID.");
+                    continue;
+                }
+            }
+        }
+    }
+}
+
+void V2XRadioClient::printVectorHelper(const std::vector<uint8_t>& vec)
+{
+    std::ostringstream oss;
+    oss << "[";
+    for (size_t i = 0; i < vec.size(); ++i) {
+        oss << static_cast<int>(vec[i]);
+        if (i < vec.size() - 1) {
+            oss << ", ";
+        }
+    }
+    oss << "]";
+    RCLCPP_DEBUG_STREAM(logger_, "Sending message vector: " << oss.str());
+}
+
+bool V2XRadioClient::isValidMsgSize(const std::vector<uint8_t> msg_vec, size_t start_index, size_t end_index, const std::vector<uint8_t> entry)
+{
+    // If message vector is larger than 255 bytes, length field will be 2 octets.
+    if (msg_vec.size() > 255) 
+    {
+        auto tmp_start_index = start_index + long_frame_;
+        std::vector<uint8_t> long_vec(entry.begin() + tmp_start_index, entry.begin() + end_index);
+        auto msg_size = (static_cast<uint16_t>(msg_vec[2]) << 8) | msg_vec[3];
+        if (msg_size == long_vec.size())
+            {
+                return true;
+            }
+        else 
+        {
+            return false;
+        }
+    }
+    // If message vector is smaller than 255 bytes, length field will be 1 octet.
+    else 
+    {
+        auto tmp_start_index = start_index + short_frame_;
+        std::vector<uint8_t> short_vec(entry.begin() + tmp_start_index, entry.begin() + end_index);
+        auto msg_size = msg_vec[2];
```

Comment:
Can we add a check first to make sure msg_vec.size() is large enough to access this element to avoid an out of range exception?

Reply:
Updated to make sure msg_vec[2] exists